lightning/src/chain/chainmonitor.rs

```diff
@@ -162,39 +184,32 @@ impl<ChanSigner: ChannelKeys, C: Deref, T: Deref, F: Deref, L: Deref> ChainMonit
 		Ok(())
 	}
 
-	/// Updates the monitor that watches the channel referred to by the given outpoint.
-	fn update_monitor(&self, outpoint: OutPoint, update: ChannelMonitorUpdate) -> Result<(), MonitorUpdateError> {
+	fn update_channel(&self, funding_txo: OutPoint, update: ChannelMonitorUpdate) -> Result<(), ChannelMonitorUpdateErr> {
+		// Update the monitor that watches the channel referred to by the given outpoint.
 		let mut monitors = self.monitors.lock().unwrap();
-		match monitors.get_mut(&outpoint) {
+		match monitors.get_mut(&funding_txo) {
+			None => {
+				log_error!(self.logger, "Failed to update channel monitor: no such monitor registered");
+				return Err(ChannelMonitorUpdateErr::PermanentFailure);
+			},
 			Some(orig_monitor) => {
 				log_trace!(self.logger, "Updating Channel Monitor for channel {}", log_funding_info!(orig_monitor));
-				orig_monitor.update_monitor(update, &self.broadcaster, &self.logger)
-			},
-			None => Err(MonitorUpdateError("No such monitor registered"))
-		}
-	}
-}
-
-impl<ChanSigner: ChannelKeys, C: Deref + Sync + Send, T: Deref + Sync + Send, F: Deref + Sync + Send, L: Deref + Sync + Send> chain::Watch for ChainMonitor<ChanSigner, C, T, F, L>
-	where C::Target: chain::Filter,
-	      T::Target: BroadcasterInterface,
-	      F::Target: FeeEstimator,
-	      L::Target: Logger,
-{
-	type Keys = ChanSigner;
-
-	fn watch_channel(&self, funding_txo: OutPoint, monitor: ChannelMonitor<ChanSigner>) -> Result<(), ChannelMonitorUpdateErr> {
-		match self.add_monitor(funding_txo, monitor) {
-			Ok(_) => Ok(()),
-			Err(_) => Err(ChannelMonitorUpdateErr::PermanentFailure),
-		}
-	}
-
-	fn update_channel(&self, funding_txo: OutPoint, update: ChannelMonitorUpdate) -> Result<(), ChannelMonitorUpdateErr> {
-		match self.update_monitor(funding_txo, update) {
-			Ok(_) => Ok(()),
-			Err(_) => Err(ChannelMonitorUpdateErr::PermanentFailure),
+				let res = orig_monitor.update_monitor(&update, &self.broadcaster, &self.logger);
+				// Even if updating the monitor returns an error, the monitor's state will
+				// still be changed. So, persist the updated monitor despite the error.
+				if let Err(e) = self.persister.update_persisted_channel(funding_txo, &update, orig_monitor) {
+					if !res.is_err() {
+						log_error!(self.logger, "Failed to persist new channel data: {:?}", e);
+						return Err(e);
+					}
+				}
+				if let Err(e) = res {
+					log_error!(self.logger, "Failed to update channel monitor: {:?}", e);
+					return Err(ChannelMonitorUpdateErr::PermanentFailure);
+				}
+			}
 		}
```

Comment:
IMO, this diff makes the code more dev-readable :) I think it doesn't change error return for failed `updated_persisted_channel`, it's already returning a `ChannelMonitorUpdateErr::PermanentFailure`

```
@@ -187,29 +188,20 @@ where C::Target: chain::Filter,
        fn update_channel(&self, funding_txo: OutPoint, update: ChannelMonitorUpdate) -> Result<(), ChannelMonitorUpdateErr> {
                // Update the monitor that watches the channel referred to by the given outpoint.
                let mut monitors = self.monitors.lock().unwrap();
-               match monitors.get_mut(&funding_txo) {
-                       None => {
-                               log_error!(self.logger, "Failed to update channel monitor: no such monitor registered");
-                               return Err(ChannelMonitorUpdateErr::PermanentFailure);
-                       },
-                       Some(orig_monitor) => {
-                               log_trace!(self.logger, "Updating Channel Monitor for channel {}", log_funding_info!(orig_monitor));
-                               let res = orig_monitor.update_monitor(&update, &self.broadcaster, &self.logger);
-                               // Even if updating the monitor returns an error, the monitor's state will
-                               // still be changed. So, persist the updated monitor despite the error.
-                               if let Err(e) = self.persister.update_persisted_channel(funding_txo, &update, orig_monitor) {
-                                       if !res.is_err() {
-                                               log_error!(self.logger, "Failed to persist new channel data: {:?}", e);
-                                               return Err(e);
-                                       }
-                               }
-                               if let Err(e) = res {
-                                       log_error!(self.logger, "Failed to update channel monitor: {:?}", e);
-                                       return Err(ChannelMonitorUpdateErr::PermanentFailure);
-                               }
+               let monitor_found = if let Some(orig_monitor) = monitors.get_mut(&funding_txo) {
+                       log_trace!(self.logger, "Updating Channel Monitor for channel {}", log_funding_info!(orig_monitor));
+                       let res = orig_monitor.update_monitor(&update, &self.broadcaster, &self.logger);
+                       // Even if updating the monitor returns an error, the monitor's state will
+                       // still be changed. So, persist the updated monitor despite the error.
+                       if let Err(e) = self.persister.update_persisted_channel(funding_txo, &update, orig_monitor) {
+                               log_error!(self.logger, "Failed to persist new channel data: {:?}", e);
+                               return Err(e);
                        }
-               }
-               Ok(())
+                       if !res.is_err() { return Ok(()) }
+                       true
+               } else { false };
+               log_error!(self.logger, "Failed to update channel monitor {}", if !monitor_found { ": no such monitor registered" } else { "" });
+               Err(ChannelMonitorUpdateErr::PermanentFailure)
        }
 
        fn release_pending_monitor_events(&self) -> Vec<MonitorEvent> {
```

Reply:
FWIW, I agree with you on the part about `updated_persisted_channel`. Though note that your suggestion results in not logging that the update failed if the persister failed. See my next comment about an alternative.

For the rest, I'd prefer the `match` over using `if let`; I think explicitly returning an error and logging each case is more readable than trying to marry the cases two as one.

Reply:
May be misunderstanding this, but the problem with changing the `update_persisted_channel` part is that we need to make sure we don't "upgrade" the error from a `PermanentFailure` to a `TemporaryFailure`, given that the `persister.update_persisted_channel` may return a temporary failure, as [Matt pointed out](https://github.com/rust-bitcoin/rust-lightning/pull/681#discussion_r494538187). & I prefer the `match` too.

Reply:
Ah, right. Do we have a test that demonstrates the intended behavior? :) This is where unit testing `ChainMonitor` with a `TestPersister` would be beneficial assuming we are able to set up a scenario that has `ChannelMonitor::update_monitor` return an error.

FWIW, this is the flattest way that I can think of representing the logic:

```diff
diff --git a/lightning/src/chain/chainmonitor.rs b/lightning/src/chain/chainmonitor.rs
index 87447f01..58f688f9 100644
--- a/lightning/src/chain/chainmonitor.rs
+++ b/lightning/src/chain/chainmonitor.rs
@@ -190,26 +190,29 @@ where C::Target: chain::Filter,
 		match monitors.get_mut(&funding_txo) {
 			None => {
 				log_error!(self.logger, "Failed to update channel monitor: no such monitor registered");
-				return Err(ChannelMonitorUpdateErr::PermanentFailure);
+				Err(ChannelMonitorUpdateErr::PermanentFailure)
 			},
 			Some(orig_monitor) => {
 				log_trace!(self.logger, "Updating Channel Monitor for channel {}", log_funding_info!(orig_monitor));
-				let res = orig_monitor.update_monitor(&update, &self.broadcaster, &self.logger);
+				let update_result = orig_monitor.update_monitor(&update, &self.broadcaster, &self.logger);
+				if let Err(e) = &update_result {
+					log_error!(self.logger, "Failed to update channel monitor: {:?}", e);
+				}
+
 				// Even if updating the monitor returns an error, the monitor's state will
 				// still be changed. So, persist the updated monitor despite the error.
-				if let Err(e) = self.persister.update_persisted_channel(funding_txo, &update, orig_monitor) {
-					if !res.is_err() {
-						log_error!(self.logger, "Failed to persist new channel data: {:?}", e);
-						return Err(e);
-					}
+				let persist_result = self.persister.update_persisted_channel(funding_txo, &update, orig_monitor);
+				if let Err(e) = &persist_result {
+					log_error!(self.logger, "Failed to persist new channel data: {:?}", e);
 				}
-				if let Err(e) = res {
-					log_error!(self.logger, "Failed to update channel monitor: {:?}", e);
-					return Err(ChannelMonitorUpdateErr::PermanentFailure);
+
+				if update_result.is_err() {
+					Err(ChannelMonitorUpdateErr::PermanentFailure)
+				} else {
+					persist_result
 				}
 			}
 		}
-		Ok(())
 	}
 
 	fn release_pending_monitor_events(&self) -> Vec<MonitorEvent> {
```

Reply:
> we need to make sure we don't "upgrade" the error from a PermanentFailure to a TemporaryFailure, given that the persister.update_persisted_channel may return a temporary failure, as Matt pointed out. 

Right, can we enforce this with the type system by defining a new struct error for persister with a method always upgrading to a PermanentFailure ? Like `PersisterError.unwrap() == ChannelMonitorUpdateErr::PermanentFailure`

Reply:
I think the scenario we're trying to avoid is returning `TemporaryFailure` from the persister when the update error is a `PermanentFailure`. It should still be possible to return a `TemporaryFailure` from the persister when the update was successful, right?

Reply:
> Do we have a test that demonstrates the intended behavior? :) 

@jkczyz took me a minute but I added this test. It's not really a unit test though, I def agree it would've been easier if we had `ChainMonitor` unit test harness things.